README.md
<h1 align="center">api-sharp</h1>

<div align="center">
<p>

[![npm](https://img.shields.io/npm/v/api-sharp)](https://www.npmjs.com/package/api-sharp) ![](https://img.shields.io/bundlephobia/minzip/api-sharp) ![](https://img.shields.io/npm/dt/api-sharp) [![CircleCI](https://img.shields.io/circleci/build/github/whinc/api-sharp/master?token=53761af868327e3798c609f9ceed6b5690147827)](https://circleci.com/dashboard) [![Coverage Status](https://coveralls.io/repos/github/whinc/api-sharp/badge.svg?branch=master)](https://coveralls.io/github/whinc/api-sharp?branch=master) [![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

</p>

<p>
跨平台的、基于配置的、轻量级的 JavaScript HTTP 客户端。
</p>

</div>

## 特性

- 基于 [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)
- 支持数据转换
- 支持自动解析 JSON
- 支持设置请求超时
- 支持缓存（内存缓存、持久化存储缓存、自定义缓存）
- 支持接口 mock
- 支持失败重试
- 支持自定义日志
- 支持 [TypeScript](http://www.typescriptlang.org/docs/home.html)
- 支持自定义 HTTP 引擎（可适配所有 JS 运行时环境）
- 支持跨平台（浏览器、node.js、小程序、ReactNative、Weex 等）
  - 浏览器使用 [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) 请求
  - node.js 使用 [http](https://nodejs.org/api/http.html) 请求
  - 即将到来...（更多平台将会内置支持，不支持的可自行实现，扩展非常简单）

| ![Chrome](https://raw.github.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png) | ![Firefox](https://raw.github.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png) | ![Safari](https://raw.github.com/alrra/browser-logos/master/src/safari/safari_48x48.png) | ![Opera](https://raw.github.com/alrra/browser-logos/master/src/opera/opera_48x48.png) | ![Edge](https://raw.github.com/alrra/browser-logos/master/src/edge/edge_48x48.png) | ![IE](https://raw.github.com/alrra/browser-logos/master/src/archive/internet-explorer_9-11/internet-explorer_9-11_48x48.png) |
| ---------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------- |
| Latest ✔                                                                                 | Latest ✔                                                                                    | Latest ✔                                                                                 | Latest ✔                                                                              | Latest ✔                                                                           | 11 ✔                                                                                                                         |

## 安装

使用 npm 安装

```bash
$ npm install api-sharp
```

使用 yarn 安装

```bash
$ yarn add api-sharp
```

## 示例

[![Edit api-sharp demo](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/api-sharp-demo-rw1n3?expanddevtools=1&fontsize=14&module=%2Fsrc%2Findex.js)

创建 ApiSharp 实例

```js
import {ApiSharp} from "api-sharp"

// 创建实例，可以传入全局配置，省略使用默认配置
const apiSharp = new ApiSharp({...})
```

发送 GET 请求

```js
// 请求服务器时间
const response = await apiSharp.request({ url: "/json/server_date" })
console.log(response)
// {
//   "data": {
//     "timestamp": 1569685733048
//   },
//   "from": "network",
//   "api": {
//     "url": "/json/server_date",
//     "method": "GET",
//     ...
//   },
//   "status": 200,
//   "statusText": "OK",
//   "headers": {
//     "content-type": "application/json",
//     "content-length": "27"
//   }
// }
```

发送 POST 请求

```js
const response = await apiSharp.request({
  url: "/json/server_date",
  method: "POST",
  body: {
    format: "json"
  }
})
```

开启缓存（仅支持 GET 请求）

```js
import {memoryCache, localStorageCache, sessionStorageCache} from "api-sharp"

const apiDescriptor = {
  url: "/json/server_date",
  enableCache: true,
  // 设置缓存时间，单位毫秒
  cacheTime: 10 * 1000,
  // 设置缓存存储位置，默认存储在内存
  cache: memoryCache
}
const response1 = await apiSharp.request(apiDescriptor)
const response2 = await apiSharp.request(apiDescriptor)
expect(response1.from).toEqual("network")
expect(response2.from).toEqual("cache")
expect(response1.data).toEqual(response2.data)
```

开启接口数据 mock

```js
const response = await apiSharp.request({
  url: "/json/server_date",
  enableMock: true,
  // 模拟数据，支持任意 JS 数据类型
  mockData: "mock data"
})
expect(response.data).toEqual("mock data")
```

开启失败重试

```js
const response = await apiSharp.request({
  url: "/json/server_date",
  enableRetry: true,
  // 最大重试次数
  retryTimes: 3
})
```

开启日志（默认开启）

```js
const response = await apiSharp.request({
  url: "/json/articles",
  method: "GET",
  description: "文章列表",
  enableLog: true,
  // 自定义日志输出
  formatLog: (type, api, _data) => {
    console.log(type, api, _data)
  }
})
```

默认日志截图：

![](./docs/screenshot1.png)

## api-sharp API

`ApiSharp`实例方法

```typescript
class ApiSharp {
  request(url: string): Promise<IResponse>
  request(api: ApiDescriptor): Promise<IResponse>
}
```

请求方法支持的接口配置项

```typescript
export type ApiDescriptor = CommonApiDescriptor & WebXhrApiDescriptor

interface CommonApiDescriptor {
  /**
   * 请求地址
   *
   * 支持相对地址（如`"/a/b/c"`）和绝对地址（如`"http://xyz.com/a/b/c"`）
   */
  url: string
  /**
   * 基地址
   *
   * 默认`""`
   *
   * 例如：`'http://xyz.com'`, `'http://xyz.com/a/b'`
   */
  baseURL?: string
  /**
   * HTTP 请求方法
   *
   * 支持 `"GET" | "POST"`
   *
   * 默认`"GET"`
   */
  method?: HttpMethod
  /**
   * HTTP 请求头
   *
   * 如果设置了全局 headers，接口中的 headers 将于全局 headers 合并，且接口中的 header 优先级更高
   *
   * 默认`{"Content-Type": "application/json"}`
   */
  headers?: HttpHeader
  /**
   * 接口描述
   *
   * 默认`""`
   */
  description?: string
  /**
   * 请求 URL 中的查询参数
   *
   * 对象会转换成 URL 查询字符串并拼接在 URL 后面，转换规则：encodeURIComponent(k1)=encodeURIComponent(v1)&encodeURIComponent(k2)=encodeURIComponent(v2)...
   *
   * 例如：`{a: 1, b: 2}`会转换成`"a=1&b=2"`
   */
  query?: QueryType
  /**
   * 请求体中的数据
   *
   * 仅支持 POST 请求，数据会转换成字符串传输，转换规则由请求头`Content-Type`决定：
   * 请求头包含`Content-Type: application/json`时，数据序列化为 JSON 字符串
   *
   * 例如：`{a: 1, b: 2}`
   */
  body?: BodyType
  /**
   * 响应的数据类型
   *
   * 支持类型如下：
   *
   * `"text"`：返回字符串
   * `"json"`：返回`JSON.parse()`后的结果，如果解析失败返回`null`
   *
   * 默认`"json"`
   */
  responseType?: ResponseType
  /**
   * 转换请求数据
   */
  transformRequest?: (body: BodyType, headers: Object) => any
  /**
   * 检查响应数据是否有效
   *
   * 检查函数返回 true 表示成功，返回 false 表示失败（失败信息为 HTTP 状态码描述)，返回 Error 也表示失败（失败信息为 Error 中的错误消息）
   *
   * 默认：`(res) => res.status >= 200 && res.status < 300`
   */
  validateResponse?: (res: IResponse) => boolean | Error
  /**
   * 转换响应数据
   */
  transformResponse?: (data: any) => any
  /**
   * 开启缓存
   *
   * 并发请求相同接口且参数相同时，实际只会发出一个请求，因为缓存的是请求的 Promise
   *
   * 默认`false`
   */
  enableCache?: boolean
  /**
   * 缓存持续时间，单位毫秒
   *
   * 下次取缓存时，如果缓存已存在的的时间超过该值，则对应缓存失效
   *
   * 默认 `5*60*1000`ms
   */
  cacheTime?: number
  /**
   * 缓存接口
   * 
   * 内置了提供基于内存、localStorage 和 sessionStorage 的缓存接口实现
   */
  cache?: ICache
  /**
   * 开启接口数据模拟
   *
   * 默认`false`
   */
  enableMock?: boolean
  /**
   * 模拟的接口数据
   *
   * 默认`undefined`
   *
   * 例如：`{id: 1, name: 'jim'}`
   */
  mockData?: any
  /**
   * 开启失败重试
   *
   * 默认`false`
   */
  enableRetry?: boolean
  /**
   * 重试最大次数
   *
   * 默认`1`
   */
  retryTimes?: number
  /**
   * 接口超时时间，单位毫秒
   *
   * 从发出请求起，如果 timeout 毫秒后接口未返回，接口调用失败。
   *
   * 默认`60*1000`ms
   */
  timeout?: number
  /**
   * 开启控制台日志
   *
   * 默认为`process.env.NODE_ENV !== "production"`
   */
  enableLog?: boolean
  /**
   * 格式化日志
   */
  formatLog?: (type: LogType, api: ProcessedApiDescriptor, data?: any) => void
}

interface WebXhrApiDescriptor {
  /**
   * 跨域请求时是否带上用户信息（如Cookie和认证的HTTP头）
   *
   * 默认`false`
   */
  withCredentials?: boolean
}
```

请求返回的数据结构

```typescript
export interface IResponse<T = any> {
  // HTTP 响应状态码
  status: number
  // HTTP 响应状态描述
  statusText: string
  // HTTP 响应数据
  data: T
  // HTTP 响应头
  headers: HttpHeader
  // 本次请求响应数据的来源
  from: "network" | "cache" | "mock"
  // 本次请求的接口描述符
  api: ProcessedApiDescriptor
}
```

## 更新日志

[CHANGELOG.md](./CHANGELOG.md)

## 贡献指南

[CONTRIBUTION.md](./CONTRIBUTION.md)

## LICENSE

[MIT](./LICENSE)


CHANGELOG.md
# [1.7.0](https://github.com/whinc/api-sharp/compare/v1.6.3...v1.7.0) (2021-03-22)


### Features

* 支持单次请求设置cache存储 ([b1720a8](https://github.com/whinc/api-sharp/commit/b1720a8))

## [1.6.3](https://github.com/whinc/api-sharp/compare/v1.6.2...v1.6.3) (2020-06-08)


### Bug Fixes

* 修复typings ([c9faa19](https://github.com/whinc/api-sharp/commit/c9faa19))

## [1.6.2](https://github.com/whinc/api-sharp/compare/v1.6.1...v1.6.2) (2020-02-23)


### Bug Fixes

* 修复types问题 ([a432d85](https://github.com/whinc/api-sharp/commit/a432d85))

## [1.6.1](https://github.com/whinc/api-sharp/compare/v1.6.0...v1.6.1) (2020-02-23)

### Bug Fixes

- 修复依赖问题和类型问题 ([47c6214](https://github.com/whinc/api-sharp/commit/47c6214))

# [1.6.0](https://github.com/whinc/api-sharp/compare/v1.5.3...v1.6.0) (2020-02-22)

### Bug Fixes

- 请求日志类型错误 ([eaa4397](https://github.com/whinc/api-sharp/commit/eaa4397))

### Features

- retryTimes 默认值改为 0 次 ([ffdb0f3](https://github.com/whinc/api-sharp/commit/ffdb0f3))
- 导出内部类型声明 ([cce70b1](https://github.com/whinc/api-sharp/commit/cce70b1))
- 导出内部类型声明 ([b120f6d](https://github.com/whinc/api-sharp/commit/b120f6d))
- 支持 node.js ([d5c90c9](https://github.com/whinc/api-sharp/commit/d5c90c9))
- 支持 node.js ([422b2b1](https://github.com/whinc/api-sharp/commit/422b2b1))
- 支持更多 HTTP 请求方法 ([ea824e9](https://github.com/whinc/api-sharp/commit/ea824e9))
- 支持更多 HTTP 请求方法 ([80c6ced](https://github.com/whinc/api-sharp/commit/80c6ced))
- 支持转换请求 ([8dd702c](https://github.com/whinc/api-sharp/commit/8dd702c))

## [1.5.3](https://github.com/whinc/api-sharp/compare/v1.5.2...v1.5.3) (2019-09-29)

### Bug Fixes

- 修复控制台错误 ([93840ce](https://github.com/whinc/api-sharp/commit/93840ce))

## [1.5.2](https://github.com/whinc/api-sharp/compare/v1.5.1...v1.5.2) (2019-09-29)

### Bug Fixes

- 修复 babel 配置错误 ([c4867f9](https://github.com/whinc/api-sharp/commit/c4867f9))

## [1.5.1](https://github.com/whinc/api-sharp/compare/v1.5.0...v1.5.1) (2019-09-28)

### Bug Fixes

- 修复测试用例 bug ([21efeee](https://github.com/whinc/api-sharp/commit/21efeee))
- 修复语法错误 ([b906cca](https://github.com/whinc/api-sharp/commit/b906cca))

# [1.5.0](https://github.com/whinc/api-sharp/compare/v1.4.0...v1.5.0) (2019-09-28)

### Bug Fixes

- 修复请求数据转化函数调用时机 ([cd02241](https://github.com/whinc/api-sharp/commit/cd02241))

### Features

- 增加请求头合并和默认请求头 ([0f8dff5](https://github.com/whinc/api-sharp/commit/0f8dff5))
- 增加请求头测试 ([aeefa3f](https://github.com/whinc/api-sharp/commit/aeefa3f))
- 支持 responseType 配置 ([c05515a](https://github.com/whinc/api-sharp/commit/c05515a))
- 新增 validateResponse 配置项 ([4225c7f](https://github.com/whinc/api-sharp/commit/4225c7f))
- 新增基于浏览器 storage 的缓存实现 ([484a58e](https://github.com/whinc/api-sharp/commit/484a58e))
- 新增配置参数 searc/seachPropTypes/body/bodyPropTypes ([b8619f7](https://github.com/whinc/api-sharp/commit/b8619f7))

# [1.4.0](https://github.com/whinc/api-sharp/compare/v1.3.2...v1.4.0) (2019-09-24)

### Features

- 浏览器平台支持 withCredentials 配置 ([2546ae1](https://github.com/whinc/api-sharp/commit/2546ae1))

## [1.3.2](https://github.com/whinc/api-sharp/compare/v1.3.1...v1.3.2) (2019-09-19)

### Bug Fixes

- 修复导出错误 ([6149301](https://github.com/whinc/api-sharp/commit/6149301))

## [1.3.1](https://github.com/whinc/api-sharp/compare/v1.3.0...v1.3.1) (2019-09-19)

### Bug Fixes

- 修复构建输出缺少 cjs 格式的问题 ([2e17a85](https://github.com/whinc/api-sharp/commit/2e17a85))
- 修复构建输出缺少 cjs 格式的问题 ([a055ee2](https://github.com/whinc/api-sharp/commit/a055ee2))


src/ApiSharp.ts
import * as PropTypes from "./PropTypes"
import { isString, getSortedString, identity, invariant, warning, isPlainObject } from "./utils"
import { ICache, MemoryCache } from "./cache"
import {
  WebXhrClient,
  IHttpClient,
  IResponse,
  HttpMethod,
  HttpHeader,
  BodyType,
  QueryType,
  ResponseType
} from "./http_client"
import { formatFullUrl } from "./utils"

const removeUndefinedValue = target => {
  return Object.keys(target)
    .filter(key => target[key] !== undefined)
    .reduce((obj, key) => Object.assign(obj, { [key]: target[key] }), {})
}

const httpMethodRegExp = /GET|POST|DELETE|HEAD|OPTIONS|PUT|PATCH/i

export interface ApiResponse<T> extends IResponse<T> {
  /**
   * 请求接口描述符
   */
  api: ProcessedApiDescriptor
  /**
   * 响应数据的来源
   */
  from: "cache" | "network" | "mock"
}

export type ApiDescriptor = CommonApiDescriptor & WebXhrApiDescriptor

interface CommonApiDescriptor {
  /**
   * 请求地址
   *
   * 支持相对地址（如`"/a/b/c"`）和绝对地址（如`"http://xyz.com/a/b/c"`）
   */
  url: string
  /**
   * 基地址
   *
   * 默认`""`
   *
   * 例如：`'http://xyz.com'`, `'http://xyz.com/a/b'`
   */
  baseURL?: string
  /**
   * HTTP 请求方法
   *
   * 支持 `GET|POST|DELETE|HEAD|OPTIONS|PUT|PATCH`
   *
   * 默认`"GET"`
   */
  method?: HttpMethod
  /**
   * HTTP 请求头
   *
   * 如果设置了全局 headers，接口中的 headers 将于全局 headers 合并，且接口中的 header 优先级更高
   *
   * 默认`{"Content-Type": "application/json"}`
   */
  headers?: HttpHeader
  /**
   * 接口描述
   *
   * 默认`""`
   */
  description?: string
  /**
   * 请求 URL 中的查询参数
   *
   * 对象会转换成 URL 查询字符串并拼接在 URL 后面，转换规则：encodeURIComponent(k1)=encodeURIComponent(v1)&encodeURIComponent(k2)=encodeURIComponent(v2)...
   *
   * 例如：`{a: 1, b: 2}`会转换成`"a=1&b=2"`
   */
  query?: QueryType
  /**
   * 请求 URL 中的查询参数类型
   *
   * 仅当 query 为`Object`类型且`process.env.NODE_ENV !== 'production'`时执行检查
   *
   * 例如：`{ id: PropTypes.number.isRequired }`
   */
  queryPropTypes?: { [key: string]: PropTypes.Validator } | null
  /**
   * 请求体中的数据
   *
   * 仅支持 POST 请求，数据会转换成字符串传输，转换规则由请求头`Content-Type`决定：
   * 请求头包含`Content-Type: application/json`时，数据序列化为 JSON 字符串
   *
   * 例如：`{a: 1, b: 2}`
   */
  body?: BodyType
  /**
   * 传入的`body`的数据类型
   *
   * 仅当 body 为`Object`类型且`process.env.NODE_ENV !== 'production'`时执行类型检查，类型检查时机发生在使用`transformRequest`进行数据转换之前
   *
   * 例如：`{ id: PropTypes.number.isRequired }`
   */
  bodyPropTypes?: { [key: string]: PropTypes.Validator } | null
  /**
   * 响应的数据类型
   *
   * 支持类型如下：
   *
   * `"text"`：返回字符串
   * `"json"`：返回`JSON.parse()`后的结果，如果解析失败返回`null`
   *
   * 默认`"json"`
   */
  responseType?: ResponseType
  /**
   * 转换请求数据
   */
  transformRequest?: (body: BodyType, headers: Object) => any
  /**
   * 检查响应数据是否有效
   *
   * 检查函数返回 true 表示成功，返回 false 表示失败（失败信息为 HTTP 状态码描述)，返回 Error 也表示失败（失败信息为 Error 中的错误消息）
   *
   * 默认：`(res) => res.status >= 200 && res.status < 300`
   */
  validateResponse?: (res: IResponse) => boolean | Error
  /**
   * 转换响应数据
   */
  transformResponse?: (data: any) => any
  /**
   * 开启缓存
   *
   * 开启后，优先返回缓存，如果无可用缓存则请求网络，并缓存返回结果
   *
   * 默认`false`
   */
  enableCache?: boolean
  /**
   * 缓存持续时间，单位毫秒
   *
   * 本次网络请求结果缓存的时间，仅当 enableCache 为 true 时有效
   *
   * 默认 `5*60*1000`ms
   */
  cacheTime?: number
  /**
   * 开启接口数据模拟
   *
   * 默认`false`
   */
  enableMock?: boolean
  /**
   * 模拟的接口数据
   *
   * 默认`undefined`
   *
   * 例如：`{id: 1, name: 'jim'}`
   */
  mockData?: any
  /**
   * 开启失败重试
   *
   * 默认`false`
   */
  enableRetry?: boolean
  /**
   * 重试最大次数
   *
   * 默认`1`
   */
  retryTimes?: number
  /**
   * 接口超时时间，单位毫秒
   *
   * 从发出请求起，如果 timeout 毫秒后接口未返回，接口调用失败。
   *
   * 默认`60*1000`ms
   */
  timeout?: number
  /**
   * 开启控制台日志
   *
   * 默认为`process.env.NODE_ENV !== "production"`
   */
  enableLog?: boolean
  /**
   * 格式化日志
   */
  formatLog?: (type: LogType, api: ProcessedApiDescriptor, data?: any) => void
}

interface WebXhrApiDescriptor {
  /**
   * 跨域请求时是否带上用户信息（如Cookie和认证的HTTP头）
   *
   * 默认`false`
   */
  withCredentials?: boolean
}

export type ProcessedApiDescriptor = Required<ApiDescriptor>

export enum LogType {
  Request,
  Response,
  ResponseError,
  ResponseCache
}

/**
 * 全局配置项
 */
export interface ApiSharpOptions extends Partial<ApiDescriptor> {
  httpClient?: IHttpClient
  cache?: ICache<IResponse>
}

const configMap = {
  [LogType.Request]: { text: "Request", bgColor: "rgba(0, 116, 217, 0.69)", fgColor: "#0074D9" },
  [LogType.Response]: { text: "Response", bgColor: "rgba(61, 153, 112, 0.69)", fgColor: "#3D9970" },
  [LogType.ResponseError]: {
    text: "Response Error",
    bgColor: "rgba(255, 65, 54, 0.69)",
    fgColor: "#FF4136"
  },
  [LogType.ResponseCache]: {
    text: "Response Cache",
    bgColor: "rgba(177, 13, 201, 0.69)",
    fgColor: "#B10DC9"
  }
}

export const defaultOptions: Required<ApiSharpOptions> = {
  httpClient: new WebXhrClient(),
  cache: new MemoryCache<IResponse>(),
  withCredentials: false,
  baseURL: "",
  headers: {
    "Content-Type": "application/json"
  },
  url: "",
  description: "",
  query: null,
  queryPropTypes: null,
  body: null,
  bodyPropTypes: null,
  enableMock: false,
  mockData: undefined,
  method: "GET",
  responseType: "json",
  enableCache: false,
  cacheTime: 5 * 60 * 1000,
  transformRequest: identity,
  transformResponse: identity,
  validateResponse: res => res.status >= 200 && res.status < 300,
  enableRetry: false,
  retryTimes: 1,
  timeout: 0,
  enableLog: process.env.NODE_ENV !== "production",
  formatLog: (type, api, data) => {
    const config = configMap[type]
    console.log(
      `%c${config.text} %c %c${api.method}|${api.description}|${api.url}%c|%O|%O`,
      `color: white; background-color: ${config.bgColor}; padding: 2px 5px; border-radius: 2px`,
      "",
      `color: ${config.fgColor}`,
      "",
      api.body,
      data
    )
  }
}

export class ApiSharp {
  private readonly options: ApiSharpOptions
  private readonly httpClient: IHttpClient
  private readonly cache: ICache<IResponse>

  constructor(options: ApiSharpOptions = {}) {
    this.options = options
    this.httpClient = options.httpClient || defaultOptions.httpClient
    this.cache = options.cache || defaultOptions.cache
  }

  /**
   * 发送 HTTP 请求
   * @param api - 接口描述符
   * @return 响应数据
   */
  async request<T = any>(_api: ApiDescriptor | string): Promise<ApiResponse<T>> {
    const api = this.processApi(_api)

    this.logRequest(api)

    // 处理 mock 数据
    if (api.enableMock) {
      return {
        data: api.mockData,
        from: "mock",
        api,
        headers: {},
        status: 200,
        statusText: "OK(mock)"
      }
    }

    let requestPromise: Promise<IResponse<T>>
    let cachedKey
    let hitCache = false

    // 处理缓存
    if (api.enableCache) {
      cachedKey = this.generateCachedKey(api)
      if (this.cache.has(cachedKey)) {
        const cachedRes = this.cache.get(cachedKey)!
        requestPromise = Promise.resolve(cachedRes)
        hitCache = true
      } else {
        requestPromise = this.sendRequest<T>(api)
      }
    } else {
      requestPromise = this.sendRequest<T>(api)
    }

    let res: IResponse<T>

    // 获取请求结果
    try {
      res = await requestPromise
    } catch (err) {
      // 处理请求异常情况

      // 请求失败或超时，都会抛出异常并被捕获处理

      // 请求失败时删除缓存
      if (api.enableCache && cachedKey) {
        this.cache.delete(cachedKey)
      }
      if (api.enableRetry && api.retryTimes >= 1) {
        return this.request({ ...api, retryTimes: api.retryTimes - 1 })
      } else {
        this.logResponseError(err, api)
        throw err
      }
    }

    // 处理请求返回情况
    const result = api.validateResponse(res)
    if (result === true) {
      // 请求成功，缓存结果（如果本次结果来自缓存，则不更新缓存，避免缓存期无限延长）
      if (api.enableCache && cachedKey && !hitCache) {
        this.cache.set(cachedKey, res, api.cacheTime)
      }
    } else {
      // 请求失败，重置缓存
      if (api.enableCache && cachedKey) {
        this.cache.delete(cachedKey)
      }
      // 失败重试
      if (api.enableRetry && api.retryTimes >= 1) {
        return this.request({ ...api, retryTimes: api.retryTimes - 1 })
      } else {
        this.logResponseError(api, res.data)
        // __DEV__ && console.error(res)
        throw result instanceof Error ? result : new Error(res.statusText)
      }
    }

    // 打印原始数据
    if (hitCache) {
      this.logResponseCache(api, res.data)
    } else {
      this.logResponse(api, res.data)
    }

    // 转换后的数据
    const transformedData = api.transformResponse(res.data)

    return {
      data: transformedData,
      from: hitCache ? "cache" : "network",
      api,
      status: res.status,
      statusText: res.statusText,
      headers: res.headers
    }
  }

  /**
   * 清除全部缓存
   */
  public clearCache() {
    return this.cache.clear()
  }

  private sendRequest<T>(api: ProcessedApiDescriptor): Promise<IResponse<T>> {
    const fullUrl = formatFullUrl(api.baseURL, api.url, api.method === "GET" ? api.query : null)
    return this.httpClient.request<T>({
      url: fullUrl,
      method: api.method,
      headers: api.headers,
      body: api.method === "POST" ? api.body : null,
      responseType: api.responseType,
      timeout: api.timeout
    })
  }

  private generateCachedKey(api: ApiDescriptor) {
    return `${api.method} ${api.baseURL}${api.url}?${getSortedString(api.query)}`
  }

  private mergeApi(
    api: ApiDescriptor,
    options: ApiSharpOptions,
    defaultOptions: Required<ApiSharpOptions>
  ): ProcessedApiDescriptor {
    const { httpClient, cache, ..._defaultOptions } = defaultOptions
    const _options = removeUndefinedValue(options) as ApiSharpOptions
    const _api = removeUndefinedValue(api) as Required<ApiSharpOptions>
    return {
      ..._defaultOptions,
      ..._options,
      ..._api,
      headers: {
        ...defaultOptions.headers,
        ..._options.headers,
        ..._api.headers
      }
    }
  }

  /**
   * 预处理接口，设置默认值、进行类型检查、数据转换等
   */
  public processApi(api: ApiDescriptor | string): ProcessedApiDescriptor {
    invariant(api, "api 为空")

    if (isString(api)) {
      api = { url: api }
    }

    const _api = this.mergeApi(api, this.options, defaultOptions)
    invariant(_api.url && isString(_api.url), "url 为空")

    _api.baseURL = _api.baseURL.replace(/\/+$/, "")

    invariant(httpMethodRegExp.test(_api.method), `无效的 HTTP 方法："${_api.method}"`)
    _api.method = _api.method.toUpperCase() as HttpMethod

    warning(
      _api.method === "GET" || !_api.enableCache,
      `只有 GET 请求支持开启缓存，当前请求方法为"${_api.method}"，缓存开启不会生效`
    )

    _api.timeout = Math.ceil(Math.max(_api.timeout, 0))

    const _query = _api.query
    // 检查查询参数类型
    if (__DEV__) {
      if (isPlainObject(_query) && isPlainObject(_api.queryPropTypes)) {
        const name = _api.baseURL + _api.url
        PropTypes.checkPropTypes(_api.queryPropTypes, _query, "", name)
      }
    }
    _api.query = _query

    const _body = _api.body
    // 检查请求数据类型
    if (__DEV__) {
      if (isPlainObject(_body) && isPlainObject(_api.bodyPropTypes)) {
        const name = _api.baseURL + _api.url
        PropTypes.checkPropTypes(_api.bodyPropTypes, _body, "", name)
      }
    }
    // 转换请求数据
    _api.body = _api.transformRequest.call(null, _body, _api.headers)

    return _api
  }

  private logRequest(api: ProcessedApiDescriptor) {
    api.enableLog && api.formatLog(LogType.Request, api)
  }

  private logResponse(api: ProcessedApiDescriptor, data) {
    api.enableLog && api.formatLog(LogType.Response, api, data)
  }

  private logResponseError(api: ProcessedApiDescriptor, data?: any) {
    api.enableLog && api.formatLog(LogType.ResponseError, api, data)
  }

  private logResponseCache(api: ProcessedApiDescriptor, data) {
    api.enableLog && api.formatLog(LogType.ResponseCache, api, data)
  }
}
